server/test/unit/paTest/handleMessageClientsTest.js

```diff
@@ -0,0 +1,268 @@
+  helpers.log(`Sent by (mock) Twilio: ${twilioMessage}`)
+  return { status: 'queued' }
+}
+
+use(sinonChai)
+
+const sandbox = sinon.createSandbox()
+
+describe('pa.js unit tests: handleMessageClients', () => {
+  beforeEach(() => {
+    sandbox.stub(db, 'getActiveClients').returns(activeClients)
+  })
+
+  afterEach(() => {
+    sandbox.restore()
+  })
+
+  describe('for a request when Twilio is operating as expected', () => {
+    beforeEach(async () => {
+      sandbox.stub(twilioHelpers, 'sendTwilioMessage').callsFake(mockSendTwilioMessage)
+
+      this.res = mockResponse(sandbox)
+      await pa.handleMessageClients(
+        {
+          body: { twilioMessage: testTwilioMessage, googleIdToken: testGoogleIdToken },
+        },
+        this.res,
+      )
+    })
+
+    it('should respond with status 200', () => {
+      expect(this.res.status).to.be.calledWith(200)
+    })
+
+    it('should respond with the Twilio message as twilioMessage', () => {
+      expect(this.res.body.twilioMessage).to.equal(testTwilioMessage)
+    })
+
+    it('should respond with the expected phone numbers in successfullyMessaged', () => {
+      const toPhoneNumbers = this.res.body.successfullyMessaged.map(twilioTraceObject => {
+        return twilioTraceObject.to
+      })
+
+      // deep equal; no other phone numbers should be contacted
+      expect(toPhoneNumbers).to.eql(successfulPhoneNumbers)
+    })
+
+    it('should respond with the expected phone numbers in failedToMessage', () => {
+      const toPhoneNumbers = this.res.body.failedToMessage.map(twilioTraceObject => {
+        return twilioTraceObject.to
+      })
+
+      // deep equal; no other phone numbers should fail
+      expect(toPhoneNumbers).to.eql(failPhoneNumbers)
+    })
+
+    it('should call twilioHelpers.sendTwilioMessage for all active clients', () => {
+      const expectedArguments = []
+
+      activeClients.forEach(client => {
+        // gather all phone numbers for this client
+        const phoneNumbers = []
+        phoneNumbers.push(...client.responderPhoneNumbers, ...client.fallbackPhoneNumbers, ...client.heartbeatPhoneNumbers)
+
+        // gather all unique phone numbers for this client
+        const uniquePhoneNumbers = new Set()
+        phoneNumbers.forEach(phoneNumber => {
+          uniquePhoneNumbers.add(phoneNumber)
+        })
+
+        // create expected arguments for each unique phone number
+        uniquePhoneNumbers.forEach(phoneNumber => {
+          expectedArguments.push({
+            toPhoneNumber: phoneNumber,
+            fromPhoneNumber: client.fromPhoneNumber,
+            twilioMessage: testTwilioMessage,
+          })
+        })
+      })
+
+      // expect twilioHelpers.sendTwilioMessage to be called with each of the expected arguments
+      expectedArguments.forEach(args => {
+        expect(twilioHelpers.sendTwilioMessage).to.be.calledWith(args.toPhoneNumber, args.fromPhoneNumber, args.twilioMessage)
```

Comment:
I think that this `forEach` loop has that multiple-expects-in-the-same-test issue.

What you could try to do is try doing the multiple comparisons on the mock's calls (using `sinon.getCall(n)` and calling `.args` on the results) and then use just a single `expect` to verify the whole lot. 

Might get a little ugly though. So maybe only worth it if you think there is value in knowing exactly which of the arguments sent and which ones didn't.

Reply:
Sounds like it would get ugly. but I'm willing to give this a shot, since I don't like the way it's implemented. Could you elaborate on how I would use `sinon.getCall(n)` in the mocks?

Reply:
Haha, ok.

So when you have a mock like `twilioHelpers.sendTwilioMessage`, you can use `.getCall(n)` to learn about the nth time it was called. So the first time, would be `twilioHelpers.sendTwilioMessage.getCall(0)` and you can get what arguments were passed into by calling `.args` on the results. So to check the array of arguments sent to the mock, you could do `twilioHelpers.sendTwilioMessage.getCall().args` and compare each of those arguments to what you would expect.

I thought that there was a way to get an array of all the calls using something like `twilioHelpers.sendTwiliomessage.calls` but I don't see it in the [documentation](https://sinonjs.org/releases/v17/spy-call/) anymore, which is too bad because then you'd be able to loop over that to get all the times it was called.

But now that I'm writing this out, I think that it would enforce an order to the calls. But in reality, we don't care if they are called in a particular order, just that they are all called exactly one time. Hmmmmm.

Reply:
That would enforce the order of the calls... In doing it this way, though, are you suggesting that I make several test cases for each call, or to do all this in one?

Reply:
I think I've come around to just leaving your test as it was to begin with. I think doing it with only a single `expect` will end up being difficult to read the output and difficult to code. So I take back my initial suggestion, but at least now you know that the `.getCall` and `args` features exist in case they come in handy elsewhere in the future.

Reply:
Yes! That feature does sound like it would come in handy in the future, so I'll keep that in my back pocket. Thanks for exploring this with me!